R/breslow.R

```diff
@@ -0,0 +1,151 @@
+  # stop("Prediction failed") # return something here?
+  rownames(surv) = 1:nrow(surv)
+  surv
+}
+
+#' @title Breslow's Cumulative Baseline Hazard
+#'
+#' @description
+#' Computes the cumulative baseline hazard using Breslow's (1972) estimator.
+#' Assumes risk predictions are derived from a **proportional hazards**
+#' regression model.
+#'
+#' @param times (`numeric()`)\cr Vector of times (train set).
+#' @param status (`numeric()`)\cr Vector of status indicators (train set).
+#' For each observation in the train set, this should be 0 (alive/censored) or
+#' 1 (dead).
+#' @param lp (`numeric()`)\cr Vector of linear predictors (train set).
+#' These are the relative score predictions from a proportional hazards model
+#' on the train set.
+#' @param eval_times (`numeric()`)\cr Vector of times to compute the cumulative
+#' baseline hazard estimations. Should be unique and increasing time points.
+#' If `NULL` (default), the unique and sorted `times` from the train set will
+#' be used.
+#'
+#' @details
+#' Breslow's approach uses a non-parametric maximum likelihood estimation of the
+#' cumulative baseline hazard function:
+#'
+#' \deqn{\hat{H}_0(t) = \sum_{i=1}^n{\frac{I(T_i \le t)\delta_i}
+#' {\sum\nolimits_{j \in R_i}e^{lp_j}}}}
+#'
+#' where:
+#' - \eqn{\hat{H}_0(t)} is the cumulative baseline hazard
+#' - \eqn{t} is the vector of time points
+#' - \eqn{n} is number of events
+#' - \eqn{T} is the vector of event times
+#' - \eqn{\delta} is the status indicator (event or censoring)
+#' - \eqn{R_i} is the risk set (number of individuals at risk just before
+#' event \eqn{i})
+#' - \eqn{lp_j} is the risk prediction (linear predictor) of individual \eqn{j}
+#' (who is part of the risk set \eqn{R_i}).
+#'
+#' For similar implementations, see `gbm::basehaz.gbm()`, `C060::basesurv()` and
+#' `xgboost.surv::sgb_bhaz()`.
+#'
+#' @returns
+#' A vector of cumulative baseline hazards, one per (increasing) time
+#' point. The times are added as names in the output vector.
+#'
+#' @references
+#' `r format_bib("cox_1972", "lin_2007")`
+#'
+#' @examples
+#' task = tsk("rats")
+#' part = partition(task, ratio = 0.8)
+#' truth_train = task$truth(part$train)
+#'
+#' learner = lrn("surv.coxph")
+#' learner$train(task, part$train)
+#' p_train = learner$predict(task, part$train)
+#'
+#' base_haz = cbhaz_breslow(times = truth_train[,1], status = truth_train[,2],
+#'                          lp = p_train$lp)
+#' head(base_haz)
+#'@export
+cbhaz_breslow = function(times = NULL, status = NULL, lp = NULL, eval_times = NULL) {
+  utimes = sort(unique(times[status == 1])) # unique, sorted event times
+  bhaz = numeric(length(utimes))
+  for(i in 1:length(utimes)) {
+    bhaz[i] = sum(times[status == 1] == utimes[i]) /
+              sum(exp(lp[times >= utimes[i]]))
+  }
+
+  eval_times = eval_times %||% sort(unique(times))
+  ok = all(diff(eval_times) > 0)
+  if (!ok) stop("'eval_times' should be monotically increasing sequence")
```

Comment:
Unneccessary assertion
```suggestion
  eval_times = sort(unique(eval_times %||% times))
```